org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/ServiceRuleTest.java

```diff
@@ -0,0 +1,182 @@
+import static org.osgi.test.common.dictionary.Dictionaries.of;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Test;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.test.common.osgi.ServiceHolder;
+import org.osgi.test.junit4.types.Foo;
+
+public class ServiceRuleTest {
+
+	@Test
+	public void basicAssumptions() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class).build()) {
+			foos.init(getClass());
+
+			assertThat(foos.isEmpty()).isTrue();
+			assertThat(foos.getTimeout()).isEqualTo(ServiceHolder.DEFAULT_TIMEOUT);
+			assertThat(foos.isRequired()).isFalse();
+		}
+	}
+
+	@Test(expected = AssertionError.class)
+	public void requiredFailsWhenNoService() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class)
+			.required()
+			.build()) {
+
+			foos.init(getClass());
+		}
+	}
+
+	@Test(expected = AssertionError.class)
+	public void requiredFailsWhenNoServiceWithTimeout() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class)
+			.requiredWithin(10)
+			.build()) {
+
+			foos.init(getClass());
+		}
+	}
+
+	@Test
+	public void successWhenService() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class).required()
+			.build(); BundleContextRule bcRule = new BundleContextRule()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					100, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			assertThat(foos.isEmpty()).isFalse();
+			assertThat(foos.size()).isEqualTo(1);
+			assertThat(foos.getTimeout()).isEqualTo(ServiceHolder.DEFAULT_TIMEOUT);
+			assertThat(foos.isRequired()).isTrue();
+		}
+	}
+
+	@Test
+	public void successWhenServiceWithTimeout() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class)
+			.requiredWithin(10)
+			.build(); BundleContextRule bcRule = new BundleContextRule()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					5, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			assertThat(foos.getService()).isNotNull();
+			assertThat(foos.getServiceReference()).isNotNull();
+			assertThat(foos.getServiceReferences()).isNotNull();
+			assertThat(foos.getServices()).isNotEmpty()
+				.contains(afoo);
+			assertThat(foos.getTimeout()).isEqualTo(10);
+			assertThat(foos.getTracked()).isNotEmpty();
+			assertThat(foos.getTrackingCount()).isEqualTo(1);
+			assertThat(foos.isRequired()).isTrue();
+			assertThat(foos.size()).isEqualTo(1);
+			assertThat(foos.isEmpty()).isFalse();
+			assertThat(foos.waitForService(20)).isNotNull();
```

Comment:
There is a balance to strike here between hard and soft assertions. But in this case, since you are making multiple tests on a single object, putting them all under a soft assert makes sense.